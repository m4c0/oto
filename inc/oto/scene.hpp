#pragma once

#include <array>
#include <optional>
#include <span>
#include <type_traits>

namespace oto {
  template<typename Tp, typename Up>
  concept same_as = std::is_same_v<Tp, Up>;

  template<typename Tp>
  concept domain = requires {
    typename Tp::actor;
    typename Tp::background;
    typename Tp::music;
    typename Tp::side;

    { Tp::actor_to_cstr(typename Tp::actor()) } -> same_as<const char *>;
    { Tp::background_to_cstr(typename Tp::background()) } -> same_as<const char *>;
    { Tp::music_to_cstr(typename Tp::music()) } -> same_as<const char *>;
    { Tp::side_to_cstr(typename Tp::side()) } -> same_as<const char *>;
  };

  template<domain D>
  struct line {
    typename D::actor actor;
    const char * text;
  };

  template<domain D>
  struct scene_meta {
    const char * name;
    std::optional<typename D::background> background;
    std::optional<typename D::music> music;
    std::optional<typename D::side> (*cast)(typename D::actor);
    std::span<const line<D>> script;
  };

  template<domain D>
  constexpr auto null_cast(typename D::side /*unused*/) {
    return std::nullopt;
  };
}
