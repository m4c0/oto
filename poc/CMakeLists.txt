function(dumper NAME)
  set(TARGET dump-${NAME})

  add_executable(${TARGET})
  target_sources(${TARGET} PRIVATE dump-${NAME}.cpp)
  target_link_libraries(${TARGET} PRIVATE oto-headers)

  add_test(NAME ${TARGET}-create COMMAND ${TARGET})
  add_test(NAME ${TARGET}-compare
    COMMAND ${CMAKE_COMMAND} -E compare_files --ignore-eol
    ${CMAKE_CURRENT_SOURCE_DIR}/${TARGET}.yaml
    ${CMAKE_CURRENT_BINARY_DIR}/${TARGET}.yaml)
  set_tests_properties(${TARGET}-compare PROPERTIES DEPENDS ${TARGET}-create)
endfunction()

dumper(action)
dumper(assets)
dumper(yaml)

add_executable(poc MACOSX_BUNDLE WIN32)
target_sources(poc PRIVATE main.cpp)
target_link_libraries(poc PRIVATE oto)

