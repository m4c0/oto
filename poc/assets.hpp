#pragma once

#include "font.hpp"
#include "main.hpp"
#include "oto/renderer.hpp"

static constexpr const auto HALF_VALUE = 128;

static void music_callback(std::span<float> /**/) {
}

static constexpr unsigned act_color(poc::domain::actor act) {
  switch (act) {
  case poc::domain::lefty:
    return 0xaa00; // NOLINT
  case poc::domain::midly:
    return 0xaa0000; // NOLINT
  case poc::domain::midly_hot:
    return 0xff0000; // NOLINT
  case poc::domain::righty:
    return 0xaa000000; // NOLINT
  }
}
static constexpr unsigned bg_color(poc::domain::background bck) {
  switch (bck) {
  case poc::domain::title:
    return 0x7700; // NOLINT
  case poc::domain::restaurant:
    return 0x770000; // NOLINT
  case poc::domain::end:
    return 0x77000000; // NOLINT
  }
}
struct asset {
  static oto::texture load_actor(poc::domain::actor act) {
    return oto::r::create_color_texture(HALF_VALUE, HALF_VALUE, act_color(act));
  }
  static oto::texture load_background(poc::domain::background bck) {
    return oto::r::create_color_texture(HALF_VALUE, HALF_VALUE, bg_color(bck));
  }
  static auto load_music(poc::domain::music /**/) {
    return music_callback;
  }
  static oto::texture load_text_background(poc::domain::background bck) {
    if (bck == poc::domain::restaurant) {
      return oto::r::create_color_texture(HALF_VALUE, HALF_VALUE, 0);
    }
    return nullptr;
  }
  static oto::texture load_text_font(poc::domain::background /**/) {
    return poc::load_text_font();
  }
  static auto rect_of_side(poc::domain::side side) {
    static constexpr const auto ACTOR_H = 400;
    static constexpr const auto ACTOR_W = 160;
    static constexpr const auto ACTOR_Y = 100;

    switch (side) {
    case poc::domain::left:
      return oto::rect { ACTOR_W * 1, ACTOR_Y, ACTOR_W, ACTOR_H };
    case poc::domain::middle:
      return oto::rect { ACTOR_W * 2, ACTOR_Y, ACTOR_W, ACTOR_H };
    case poc::domain::right:
      return oto::rect { ACTOR_W * 3, ACTOR_Y, ACTOR_W, ACTOR_H };
    }
  }
  static auto size_of_font(poc::domain::background /**/) {
    return oto::size { .w = 16, .h = 32 };
  }
};
