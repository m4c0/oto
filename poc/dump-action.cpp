#include "dumper.hpp"
#include "oto/dumper.action.hpp"

template<typename D>
struct dumper : public oto::dumper::action<D> {
  using oto::dumper::action<D>::action;

  const oto::transition<D> * choose(oto::opcodes::choose<D> c) override {
    return &c[1].next;
  }
};

int main(int /*argc*/, char ** argv) {
  return poc::tool<dumper>(*argv);
}
