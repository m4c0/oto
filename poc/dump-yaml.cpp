#include "dumper.hpp"
#include "oto/dumper.yaml.hpp"

int main(int /*argc*/, char ** argv) {
  return poc::tool<oto::dumper::yaml>(*argv);
}
