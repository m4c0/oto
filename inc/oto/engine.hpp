#pragma once

#include "oto/assets.hpp"
#include "oto/renderer.hpp"
#include "oto/scene.hpp"
#include "oto/v_engine.hpp"
#include "oto/vm.hpp"

namespace oto {
  template<domain D, assets<D> A>
  class engine : public v_engine {
    enum state { run, pause, sleep, speak };

    texture m_actor {};
    texture m_background {};
    std::chrono::time_point<clock> m_timer {};
    state m_state = run;

    vm<D> m_vm;

    state step() {
      switch (m_state) {
      case run:
        return m_vm ? std::visit(*this, m_vm.iterate()) : pause;
      case sleep:
        return (clock::now() > m_timer) ? run : sleep;
      default:
        return m_state;
      }
    }

  public:
    explicit engine(const transition<D> * init) : m_vm(init) {
    }

    state operator()(const opcodes::background<D> & bck) {
      m_background = A::load_background(*bck);
      return run;
    }
    state operator()(const opcodes::choose<D> & /**/) {
      return pause;
    }
    state operator()(const opcodes::music<D> & mus) {
      oto::r::set_audio_callback(A::load_music(*mus));
      return run;
    }
    state operator()(const opcodes::pause & /**/) {
      return pause;
    }
    state operator()(const opcodes::present & /**/) {
      using namespace std::literals;
      m_timer = clock::now() + 2s;
      return sleep;
    }
    state operator()(const opcodes::speak<D> & spk) {
      m_actor = A::load_actor(spk.actor);
      return speak;
    }
    state operator()(std::monostate /**/) {
      return pause;
    }

    void key_down() override {
      if (m_state == pause) {
        m_state = run;
        m_actor.reset();
      }
      if (m_state == speak) m_state = run;
    }

    void run_frame() override {
      do {
        m_state = step();
      } while (m_state == run);

      oto::r::prepare();
      if (m_background) oto::r::draw(m_background);
      if (m_actor) oto::r::draw(m_actor);
      oto::r::present();
    }
  };
}
