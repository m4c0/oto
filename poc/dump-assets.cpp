#include "dumper.hpp"
#include "oto/dumper.assets.hpp"

int main(int /*argc*/, char ** argv) {
  return poc::tool<oto::dumper::assets>(*argv);
}
