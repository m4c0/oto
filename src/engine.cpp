#include "oto/engine.hpp"

#include <SDL.h>
#include <stdexcept>

static void init_audio() {
  if (SDL_Init(SDL_INIT_AUDIO) != 0) {
    throw std::runtime_error("Failed to initialise SDL audio");
  }

  static constexpr const SDL_AudioSpec desired {
    .freq = 44100,
    .format = AUDIO_F32,
    .channels = 1,
    .samples = 4096,
    //.callback = audio_callback,
  };
  SDL_AudioSpec obtained;
  auto dev = SDL_OpenAudioDevice(nullptr, 0, &desired, &obtained, 0);

  SDL_PauseAudioDevice(dev, 0);
}

oto::v_engine::v_engine(void * rnd) : m_renderer(static_cast<SDL_Renderer *>(rnd)) {
  init_audio();
}

oto::v_engine::~v_engine() {
  SDL_DestroyTexture(m_background);
}

void oto::v_engine::repaint() {
  static constexpr const auto FULL_BRIGHT = 255;
  SDL_SetRenderDrawColor(m_renderer, FULL_BRIGHT, 0, FULL_BRIGHT, FULL_BRIGHT);
  SDL_RenderClear(m_renderer);

  SDL_RenderCopy(m_renderer, m_background, nullptr, nullptr);

  SDL_RenderPresent(m_renderer);
}
