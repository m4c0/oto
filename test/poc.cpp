#include "oto/engine.hpp"
#include "poc.hpp"

static constexpr const auto bg_color(poc::domain::background bck) {
  switch (bck) {
  case poc::domain::title:
    return 0x7700;
  case poc::domain::restaurant:
    return 0x770000;
  case poc::domain::end:
    return 0x77000000;
  }
}

class poc_engine : public oto::engine<poc::domain> {
protected:
  oto::texture load_background(poc::domain::background bck) override {
    return oto::r::create_color_texture(128, 128, bg_color(bck));
  }

public:
  explicit poc_engine() : engine(&poc::game) {
  }
};

oto::v_engine * oto::create_engine() {
  return new poc_engine();
}
