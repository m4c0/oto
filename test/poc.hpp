#pragma once

#include "oto/scene.hpp"

namespace poc {
  struct domain {
    enum actor { lefty, midly, midly_hot, righty };
    enum background { title, restaurant, end };
    enum music { romance, game_over };
    enum side { left, middle, right };

    [[nodiscard]] static constexpr const char * actor_to_cstr(actor a) {
      switch (a) {
      case lefty:
        return "lefty";
      case midly:
        return "midly";
      case midly_hot:
        return "midly-hot";
      case righty:
        return "righty";
      }
    }
    [[nodiscard]] static constexpr const char * background_to_cstr(background b) {
      switch (b) {
      case title:
        return "title";
      case restaurant:
        return "restaurant";
      case end:
        return "end";
      }
    }
    [[nodiscard]] static constexpr const char * music_to_cstr(music m) {
      switch (m) {
      case romance:
        return "romance-1";
      case game_over:
        return "game-over";
      }
    }
    [[nodiscard]] static constexpr const char * side_to_cstr(side s) {
      switch (s) {
      case left:
        return "left";
      case middle:
        return "middle";
      case right:
        return "right";
      }
    }
  };
  using lin = oto::line<domain>;
  using osd = std::optional<domain::side>;
  using scm = oto::scene_meta<domain>;

  static constexpr const auto title_meta = scm {
    .name = "Title",
    .background = domain::title,
    .music = domain::romance,
  };
  static constexpr const auto lefty_meta = scm {
    .name = "Lefty Intro",
  };
  static constexpr const auto midly_script = std::array { lin { domain::midly, "This is me" } };
  static constexpr const auto midly_meta = scm {
    .name = "Midly Intro",
    .cast = [](auto /*a*/) -> osd {
      return { domain::left };
    },
    .script = std::span { midly_script },
  };
  static constexpr const auto righty_meta = scm {
    .name = "Righty Intro",
  };

  static constexpr const auto the_end_meta = scm {
    .name = "The End",
    .background = domain::end,
    .music = domain::game_over,
  };

  [[nodiscard]] static constexpr auto restaurant_meta(scm base) {
    base.background = domain::restaurant;
    base.music = domain::romance;
    return base;
  }
  static constexpr const auto intro_script = std::array {
    lin { domain::lefty, "Hello! How are you doing?" },
    lin { domain::lefty, "Do you want to know what's happening?" },
    lin { domain::righty, "Yeah, you! Baby, baby!" },
    lin { domain::midly, "Click one option to select who you want to date" },
    lin { domain::midly_hot, "I'm as hot as an idol" },
  };
  static constexpr const auto intro_meta = restaurant_meta(scm {
      .name = "Intro",
      .cast = [](domain::actor a) -> osd {
        switch (a) {
        case domain::lefty:
          return { domain::left };
        case domain::midly:
        case domain::midly_hot:
          return { domain::middle };
        case domain::righty:
          return { domain::right };
        }
      },
      .script = intro_script,
  });

  //// Transitions ////

  using edg = oto::edge<domain>;
  using trn = oto::transition<domain>;

  static constexpr const trn the_end = { .from = the_end_meta };
  static constexpr const auto end = std::array { edg { "", the_end } };

  static constexpr const trn lefty = { .from = lefty_meta, .choices = end };
  static constexpr const trn midly = { .from = midly_meta, .choices = end };
  static constexpr const trn righty = { .from = righty_meta, .choices = end };

  static constexpr const auto intro_choices = std::array {
    edg { "Lefty", lefty },
    edg { "Midly", midly },
    edg { "Righty", righty },
  };
  static constexpr const trn intro = {
    .from = intro_meta,
    .choices = { intro_choices },
  };

  static constexpr const auto title_choices = std::array {
    edg { nullptr, intro },
  };
  static constexpr const trn title = {
    .from = title_meta,
    .choices = title_choices,
  };

  static constexpr const auto game = title;
}
