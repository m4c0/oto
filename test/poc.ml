let title_meta = Oto.scene_meta ~background:"title" ~music:"romance-1" "Title"
let lefty_meta = Oto.scene_meta "Lefty Intro"
let midly_meta = Oto.scene_meta "Midly Intro"
let righty_meta = Oto.scene_meta "Righty Intro"
let lefty_says = Oto.speak "lefty"
let midly_says = Oto.speak "midly"
let righty_says = Oto.speak "righty"
let midly_poses = Oto.pose "midly"
let restaurant_meta = Oto.scene_meta ~background:"restaurant" ~music:"romance-1"

let intro_meta =
  restaurant_meta "Intro"
    ~actors:(Oto.simple_cast ~left:"lefty" ~middle:"middly" ~right:"righty")
    ~script:
      [
        lefty_says "Hello! How are you doing?";
        lefty_says "Do you want to know what's happening?";
        righty_says "Yeah, you! Baby, baby!";
        midly_says "Click one option to select who you want to date";
        midly_poses "imponent";
        midly_says "I'm as hot as an idol";
      ]

let lefty _ = (lefty_meta, Oto.EndGame)
let midly _ = (midly_meta, Oto.EndGame)
let righty _ = (righty_meta, Oto.EndGame)

let intro _ =
  ( intro_meta,
    Oto.Choice [ ("Lefty", lefty); ("Midly", midly); ("Righty", righty) ] )

let game : Oto.scene = fun _ -> (title_meta, Oto.Continuation intro)

let rec to_yaml (fn : Oto.scene) =
  let meta, next = fn () in
  print_string "name: ";
  print_endline meta.name;
  print_endline "---";
  match next with
  | Continuation s -> to_yaml s
  | WaitThenJump s -> to_yaml s
  | Choice l -> List.iter (fun (_, s) -> to_yaml s) l
  | EndGame -> ()

let () = to_yaml game
