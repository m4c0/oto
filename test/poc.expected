name: Title
---
name: Intro
---
name: Lefty Intro
---
name: Midly Intro
---
name: Righty Intro
---
